ffi/src/scan.rs

```diff
@@ -147,7 +159,8 @@ pub struct KernelScanDataIterator {
     engine: Arc<dyn ExternEngine>,
 }
 
-impl BoxHandle for KernelScanDataIterator {}
+#[handle_descriptor(target=KernelScanDataIterator, mutable=true, sized=true)]
+pub struct KernelScanDataIteratorHandle;
```

Comment:
Rescuing https://github.com/delta-incubator/delta-kernel-rs/pull/182#discussion_r1603864171

> remove Handle from name

This is a conflict between the current cbindgen capability vs. what cbindgen "should" be, combined with name mangling for C:
* Today -- cbindgen exposes `Handle<T>` in C++ as `template <typename T> typedef Handle<T> = T`, but any given `Handle<FooHandle>` will mangle to `HandleFooHandle` in the resulting C header. Which is indeed annoying. 
* Some day (ideally), cbindgen would understand that `Handle<T>` is really just a `T*` and would expose it as such. The resulting header would just have `FooHandle*` everywhere.

However, even today both C and C++ can still use a simple `FooHandle*` -- and probably _should_ do so, since the typedef doesn't add anything useful (we already have the indirection and opacity we desire from the handle itself). The one exception is things like `ExternResult<Handle<FooHandle>>` which with today's cbindgen would continue mangling badly. And since it's not a typedef we can't just ignore it (have to actually fix cbindgen). 

I guess the other approach we could do is give the kernel types the same names as their corresponding FFI types? Something like:
```rust
use delta_kernel::{Snapshot => KernelSnapshot};

#[handle_descriptor(target=KernelSnapshot, mutable=true, sized=true)]
pub struct Snapshot;
```
Downside is, that only has type safety on the rust side; on the C/C++ side one could be tempted to cast `delta_kernel_ffi::Snapshot` as `delta_kernel::Snapshot`, since they're the same name. 

Then again, if somebody's really determined to do shady casting of a type we've officially defined as opaque, then they probably deserve whatever bugs they cause.

Reply:
Yeah, I think this is okay for now. I prefer to keep the ffi types distinct from the kernel types, just to avoid confusion.

Changes to cbindgen that would be nice:
- rename on a generic struct renames on the mangled name. So a rename rule of `ExternResult` -> `Result` would let us have things like `ResultFooHandle`, since it's not really "extern" for the engine
- configure the mangling to put the outer type at the end, so we could get `FooHandleResult` which is probably the most "correct" name

Reply:
As a side effect of fixing https://github.com/delta-incubator/delta-kernel-rs/pull/182#discussion_r1604079239, all handles became shared except the one for `EngineData`, and so could be renamed as `SharedXxx`. The current code for engine data handles is illegal/incorrect, and will likely change, so for now I just imported the kernel `EngineData` as `KernelEngineData` so the FFI one could be called `EngineData`.